### ISSUE TITLE
Custom cell metadata isn't saved to disk.

### ISSUE THREAD

**Issue by brettfo** [microsoft/vscode-jupyter#5477]
## Environment data

Version: 1.56.0-insider (user setup)
Commit: 845efe5d8e31079cc858131c95c36103d9439e74
Date: 2021-04-09T09:50:06.816Z
Electron: 11.4.2
Chrome: 87.0.4280.141
Node.js: 12.18.3
V8: 8.7.220.31-electron.0
OS: Windows_NT x64 10.0.19042

Jupyter extension version: v2021.6.732501275

## Expected behaviour

1. Install the .NET Interactive extension.
2. Create a new notebook and select the .NET Interactive kernel and ensure the cell's language is "C# (.NET Interactive)"
3. Save the notebook.
4. Open in notepad.

Expected:
Cell metadata should look like

``` json
"metadata": {
  "dotnet_interactive": {
    "language": "csharp"
  }
}
```

## Actual behaviour

Cell metadata is empty.  We're updating the metadata [here](https://github.com/dotnet/interactive/blob/main/src/dotnet-interactive-vscode/common/vscode/extension.ts#L202-L211) that makes the object look like:

``` json
"metadata": {
  "custom": {
    "dotnet_interactive": {
      "language": "csharp"
    }
  }
}
```

But it's not getting persisted to disk.

**Comment by joyceerhl** [comment]
I debugged our extension locally and created a new blank notebook then selected the .NET interactive kernel, and it seems like although we have registered a handler for `vscode.notebook.onDidChangeCellMetadata`, our handler never gets called (whereas our handlers for `vscode.notebook.onDidChangeNotebookCells` and `vscode.notebook.onDidChangeNotebookDocumentMetadata` do get called):

https://github.com/microsoft/vscode-jupyter/blob/4d854f28adb60b5af0193eeb2a2b466df8858c1a/src/client/common/application/notebook.ts#L115-L117 

This suggests at least two possibilities: VS Code isn't firing the metadata change event, or perhaps the metadata edit is not actually applied to the notebook document. 

@brettfo I'm not familiar with the dotnet interactive source code so I have a question. Based on the dotnet interactive source code, it seems like [`updateNotebookCellLanguageInMetadata`](https://github.com/dotnet/interactive/blob/9d658bd6e73fb050f2d7c5edeba6bfd8ac519f6c/src/dotnet-interactive-vscode/common/vscode/extension.ts#L197 ) is only called once, when the notebook document is [first opened](https://github.com/dotnet/interactive/blob/e73aa2ef10ef6691684397feed9da27b8008d63d/src/dotnet-interactive-vscode/common/vscode/extension.ts#L175). Since setting the .NET interactive language kernel happens after the document is opened, is it also necessary to call `updateNotebookCellLanguageInMetadata` in response to something like `vscode.notebook.onDidChangeActiveKernel` in order for the metadata update to take?

**Comment by DonJayamanne** [comment]
Oh wait, I don't think we support this today.

**Comment by colombod** [comment]
This is a big issue as every time a notebook is open the user needs to reselect all languages correctly. It is impacting a lot our user experience, @DonJayamanne  do you have any sense of the work plan for this?

**Comment by nhirschey** [comment]
@colombod thank you for pushing this issue as a priority with the vscode-jupyter team. 

I traced the problem described in this comment https://github.com/dotnet/interactive/issues/1121#issuecomment-818620588 to the present issue about cell metadata not being persisted to disk.

To reiterate the point about the impact of this issue "in the wild": I had a lot of success with .net notebooks this fall and winter. But this issue makes .net interactive notebooks unusable for me.

**Comment by colombod** [comment]
> @colombod thank you for pushing this issue as a priority with the vscode-jupyter team.
> 
> I traced the problem described in this comment [dotnet/interactive#1121 (comment)](https://github.com/dotnet/interactive/issues/1121#issuecomment-818620588) to the present issue about cell metadata not being persisted to disk.
> 
> To reiterate the point about the impact of this issue "in the wild": I had a lot of success with .net notebooks this fall and winter. But this issue makes .net interactive notebooks unusable for me.

Thank you for the feedback

**Comment by colombod** [comment]
User @pkese is also reporting the problem, just mentioning the issue reported (dotnet/interactive#1266) so it can be tracked by the user too.

**Comment by colombod** [comment]
@DonJayamanne is this issue still planned for the [May Release milestone](https://github.com/microsoft/vscode-jupyter/milestone/7)?

**Comment by DonJayamanne** [comment]
Yes

**Comment by colombod** [comment]
> Yes

What should we do in the meanwhile? This is affecting all our users pretty significantly, do we need to point them at alternative file formats?

**Comment by DonJayamanne** [comment]
> do we need to point them at alternative file formats?

Sorry at couldn't get to this, I'll look into this first thing tomorrow (was busy looking at the heading api changes from vscode).
Fyi, I couldn't just the net extension in insiders, to fast this earlier today. Hopefully it's fixed.

Fyi, PRs are welcome too, as alternative to looking at other formats.

**Comment by DonJayamanne** [comment]
@colombod @brettfo This is a blocker for me in VSCode Stable https://github.com/dotnet/interactive/issues/1293

**Comment by brettfo** [comment]
Added a comment to the issue you linked with unblocking instructions.

**Comment by DonJayamanne** [comment]
@brettfo The problem is because the structure is not what we expect:
This is what it needs to be.
It shoudl be `custom.metadata`
THis is because we support attachments & other cell cell metadata
E.g. `custom.attachments` (again part of Jupyter cell level information).

```json
"metadata": {
  "custom": {
	"metadata": {
	    "dotnet_interactive": {
    	  "language": "csharp"
	    }
	 }
  }
}
```

**Comment by DonJayamanne** [comment]
Changing that should address this issue.
Closing this issue.
FYI - Looks like .NET extnsion is blowing away the existing metadata. It needs to append the metadata.
E.g. ifi I have some metadata in there, it should not blow it away
I think it will get blown away (as the .NET ipynb does't prerve any existing cell metadata)

Closing this as the structure isn't correct.

**Comment by DonJayamanne** [comment]
If that doesn't work, please feel free to re-open or file a new issue.
